system/include/emscripten/emscripten.h

```diff
@@ -255,6 +255,43 @@ typedef void (*em_scan_func)(void*, void*);
 void emscripten_scan_registers(em_scan_func func);
 void emscripten_scan_stack(em_scan_func func);
 
+// Fibers API
+// Requires upstream backend and Asyncify
+
+#define EM_FIBER_ASYNCIFY_STACK_SIZE 4096
+
+typedef struct asyncify_fiber_s {
+    void *stack_base;
+    void *stack_limit;
+    void *stack_ptr;
+    em_arg_callback_func entry;
+    void *user_data;
+} asyncify_fiber_t;
+
+typedef struct asyncify_data_s {
+    void *stack_ptr;
+    void *stack_limit;
+    int rewind_id;
+} asyncify_data_t;
```

Comment:
Can you explain why need need three different structs here?

Also, I'm a little confused by using `asyncify` in the naming.  I would have though asyncify would be an implementation detail of fibre API so this seems a little leaky.

Reply:
> Can you explain why need need three different structs here?

It's mostly for logical organization purposes: `asyncify_fiber_t` describes fiber-specific state, `asyncify_data_t` is a general Asyncify data structure used for the unwind/rewind mechanism, and `emscripten_fiber_t` is a convenience that binds them together with a default Asyncify call stack size.

> Also, I'm a little confused by using asyncify in the naming. I would have though asyncify would be an implementation detail of fibre API so this seems a little leaky.

It is kind of deliberately leaky. The idea here is that the user can either ignore all of this and simply use the provided init API to initialize the fiber struct, or dig into the implementation details if they want to squeeze a little more performance out of it. That way they can initialize the struct without an extra call into JS, which is nice for realtime applications that often spin up short-lived coroutines. There is a full initialization example in the test case, and koishi has a [partial initialization example](https://github.com/taisei-project/koishi/blob/26216c780b13279f55812554d9f6bee1ec498ee5/src/emscripten/emscripten.c#L56) suitable for re-using fibers.